File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/app/interfaces/BaseApp.java
package de.marmaro.krt.ffupdater.app.interfaces;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;

import androidx.preference.PreferenceManager;

import org.apache.commons.codec.binary.ApacheCodecHex;

import java.util.Optional;

import de.marmaro.krt.ffupdater.app.interfaces.Compatibility;
import de.marmaro.krt.ffupdater.app.interfaces.Display;
import de.marmaro.krt.ffupdater.app.interfaces.InstallerInfo;
import de.marmaro.krt.ffupdater.app.interfaces.UpdateCheck;

public abstract class BaseApp implements Display, InstallerInfo, Compatibility, UpdateCheck {

    @Override
    public boolean isInstalled(Context context) {
        try {
            context.getPackageManager().getPackageInfo(getPackageName(), 0);
            return true;
        } catch (PackageManager.NameNotFoundException e) {
            return false;
        }
    }

    @Override
    public byte[] getSignatureHash() {
        return ApacheCodecHex.decodeHex(getSignatureHashAsString());
    }

    protected Optional<String> getInstalledVersionFromPackageManager(Context context) {
        try {
            final PackageManager pm = context.getPackageManager();
            final PackageInfo packageInfo = pm.getPackageInfo(getPackageName(), 0);
            return Optional.of(packageInfo.versionName);
        } catch (PackageManager.NameNotFoundException e) {
            return Optional.empty();
        }
    }

    protected Optional<String> getInstalledVersionFromSharedPreferences(Context context, String key) {
        final SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        return Optional.ofNullable(preferences.getString(key, null));
    }

    protected void setInstalledVersionInSharedPreferences(Context context, String key, String value) {
        final SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        preferences.edit()
                .putString(key, value)
                .apply();
    }
}

File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/app/BaseApp.kt
package de.marmaro.krt.ffupdater.app

import android.content.Context
import androidx.annotation.AnyThread
import androidx.annotation.MainThread
import androidx.preference.PreferenceManager
import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import de.marmaro.krt.ffupdater.R
import de.marmaro.krt.ffupdater.device.ABI
import de.marmaro.krt.ffupdater.download.PackageManagerUtil
import de.marmaro.krt.ffupdater.security.FingerprintValidator
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.io.File


abstract class BaseApp {
    abstract val packageName: String
    abstract val displayTitle: Int
    abstract val displayDescription: Int
    abstract val displayWarning: Int?
    abstract val displayDownloadSource: Int
    abstract val displayIcon: Int
    abstract val minApiLevel: Int
    abstract val supportedAbis: List<ABI>
    abstract val signatureHash: String

    // The installation does not require special actions like rooting the smartphone
    abstract val normalInstallation: Boolean

    private val mutex = Mutex()

    @AnyThread
    fun isInstalled(context: Context): Boolean {
        return PackageManagerUtil(context.packageManager).isAppInstalled(packageName) &&
                FingerprintValidator(context.packageManager).checkInstalledApp(this).isValid
    }

    @AnyThread
    open fun getInstalledVersion(context: Context): String? {
        return PackageManagerUtil(context.packageManager).getInstalledAppVersionName(packageName)
    }

    @AnyThread
    fun getDisplayInstalledVersion(context: Context): String {
        return context.getString(R.string.installed_version, getInstalledVersion(context))
    }

    @AnyThread
    fun getDisplayAvailableVersion(context: Context, availableVersionResult: AvailableVersionResult): String {
        return context.getString(R.string.available_version, availableVersionResult.version)
    }

    @AnyThread
    open fun appIsInstalled(context: Context, available: AvailableVersionResult) {
    }

    suspend fun checkForUpdateAsync(context: Context): Deferred<UpdateCheckResult> {
        return withContext(Dispatchers.IO) {
            async {
                // - use mutex lock to prevent multiple simultaneously update check for a single app
                mutex.withLock {
                    checkForUpdateAndCacheIt(context, true)
                }
            }
        }
    }

    suspend fun checkForUpdateWithoutCacheAsync(context: Context): Deferred<UpdateCheckResult> {
        return withContext(Dispatchers.IO) {
            async {
                // - use mutex lock to prevent multiple simultaneously update check for a single app
                mutex.withLock {
                    checkForUpdateAndCacheIt(context, false)
                }
            }
        }
    }

    private suspend fun checkForUpdateAndCacheIt(context: Context, useCache: Boolean): UpdateCheckResult {
        val cacheKey = "cached_update_check_result__${packageName}"
        val preferences = PreferenceManager.getDefaultSharedPreferences(context)
        try {
            if (useCache) {
                preferences.getString(cacheKey, null)
                    ?.let { gson.fromJson(it, UpdateCheckResult::class.java) }
                    ?.takeIf { System.currentTimeMillis() - it.timestamp <= CACHE_TIME }
                    ?.let { return it }
            }
        } catch (e: JsonSyntaxException) {
            // just ignore the invalid cache
        }

        val available = checkForUpdate()
        val result = UpdateCheckResult(
            availableResult = available,
            isUpdateAvailable = isAvailableVersionHigherThanInstalled(context, available),
            displayVersion = getDisplayAvailableVersion(context, available)
        )

        preferences.edit().putString(cacheKey, gson.toJson(result)).apply()
        return result
    }

    @MainThread
    protected abstract suspend fun checkForUpdate(): AvailableVersionResult

    @MainThread
    open suspend fun isAvailableVersionEqualToArchive(
        context: Context,
        file: File,
        available: AvailableVersionResult
    ): Boolean {
        val archiveVersion = PackageManagerUtil(context.packageManager)
            .getPackageArchiveVersionNameOrNull(file.absolutePath) ?: return false
        return VersionCompareHelper.isAvailableVersionEqual(archiveVersion, available.version)
    }

    @AnyThread
    open fun isAvailableVersionHigherThanInstalled(
        context: Context,
        available: AvailableVersionResult
    ): Boolean {
        val installedVersion = getInstalledVersion(context) ?: return true
        return VersionCompareHelper.isAvailableVersionHigher(installedVersion, available.version)
    }

    companion object {
        const val CACHE_TIME = 600_000L // 10 minutes
        val gson = Gson()
    }
}